### PR TITLE
feat(downloader): error out if app receives unexpected HTTP responses

### LINKED ISSUE #77: Downloader should error out when downloaded file type is not music
Multiple ways to achieve this : Size check, if it's less than half a meg, highly probable that it's not music.

Better solution would be to check the "content-type" header of the server's response.

### PR DESCRIPTION
**Pull request by defvs**
Fixes #77 
Still WIP because didn't find a way to test yet.

### CODE REVIEW

File: src/main/xerus/monstercat/downloader/Download.kt [hunk 1]
```diff
@@ -81,10 +81,15 @@ class ReleaseDownload(private val release: Release, private var tracks: Collecti
 	
 	fun downloadTrack(releaseId: String, trackId: String, path: Path) {
 		val connection = APIConnection("api", "release", releaseId, "download").addQueries("method=download", "type=${QUALITY()}", "track=$trackId")
-		val entity = connection.getResponse().entity
+		val httpResponse = connection.getResponse()
+		val entity = httpResponse.entity
 		val contentLength = entity.contentLength
 		if(contentLength == 0L)
 			throw EmptyResponseException(connection.uri.toString())
+		if(!listOf("application/octet-stream", "audio/flac", "audio/wav", "audio/mpeg").contains(entity.contentType.value))
```
Comment (Xerus2000) [inline src/main/xerus/monstercat/downloader/Download.kt]: I have a problem with this list - what if they offer ogg or some other format in the future? I'd rather blacklist than whitelist.
Or simply check if it starts with some sort of multimedia-prefix.

File: src/main/xerus/monstercat/downloader/Download.kt [hunk 2]
```diff
@@ -81,10 +81,15 @@ class ReleaseDownload(private val release: Release, private var tracks: Collecti
...
+			throw WrongResponseTypeException(connection.uri.toString(), entity.contentType.value)
+		if(httpResponse.statusLine.statusCode != 200)
```
Comment (Xerus2000) [inline src/main/xerus/monstercat/downloader/Download.kt]: All 2xx codes indicate success, not just 200

File: src/main/xerus/monstercat/downloader/Download.kt [hunk 3]
```diff
@@ -147,3 +152,7 @@ class ReleaseDownload(private val release: Release, private var tracks: Collecti
 }
 
 class EmptyResponseException(term: String) : Exception("No file found for $term!")
+
+class WrongResponseTypeException(term: String, mime: String) : Exception("Wrong MIME type returned for $term: type $mime unexpected!")
+
+class WrongResponseCodeException(term: String, code: Int) : Exception("Response code is not \"200 OK\" for $term : is $code")
```
Comment (Xerus2000) [inline src/main/xerus/monstercat/downloader/Download.kt]: I'd reword this to make it more understandable to the average user, something like this:
```suggestion
class WrongResponseCodeException(term: String, code: Int) : Exception("Error downloading $term: Server returned code $code")
```
and maybe include the full status line, not just the code

File: src/main/xerus/monstercat/downloader/Download.kt (hunk 2 above)
Comment (defvs) [reply src/main/xerus/monstercat/downloader/Download.kt]: But some 2xx codes are "resource created" which could be harmful. 200 is THE "GOOD DOWNLOAD" one.

File: src/main/xerus/monstercat/downloader/Download.kt (hunk 1 above)
Comment (defvs) [reply src/main/xerus/monstercat/downloader/Download.kt]: I'll add the ``audio`` prefix

File: src/main/xerus/monstercat/downloader/Download.kt (hunk 3 above)
Comment (defvs) [reply src/main/xerus/monstercat/downloader/Download.kt]: Sure, I'll add the status text as well.

File: src/main/xerus/monstercat/downloader/Download.kt (hunk 1 above)
Comment (defvs) [reply src/main/xerus/monstercat/downloader/Download.kt]: 630987bcf7bbd3769fbd774023501f9d272df537 is let ok or is run better ?

File: src/main/xerus/monstercat/downloader/Download.kt (hunk 3 above)
Comment (defvs) [reply src/main/xerus/monstercat/downloader/Download.kt]: 6c9dd3d3da212c2e79512595a7811052f37bc0f9

Comment (defvs) [reply src/main/xerus/monstercat/downloader/Download.kt]: 6c9dd3d3da212c2e79512595a7811052f37bc0f9

File: src/main/xerus/monstercat/downloader/Download.kt (hunk 1 above)
Comment (Xerus2000) [reply src/main/xerus/monstercat/downloader/Download.kt]: unless there is a significant advantage, you should always prefer to have the receiver as parameter rather than receiver, so `let` over `run`. So that's correct 👍

**Comment by Xerus2000** [comment]
Unless Monstercat has fixed it, downloading any of these as FLAC should error out
https://cdn.discordapp.com/attachments/386978009741131787/523403212875759628/broken-flacs.txt
So you can use that to test.

**Comment by defvs** [comment]
They fixed it, I've already all the FLACs that CyanoTex told us didn't work.
